lightning/src/ln/peers/encryption.rs

```diff
@@ -0,0 +1,414 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Handles all over the wire message encryption and decryption upon handshake completion.
+
+use ln::peers::{chacha, hkdf5869rfc};
+use util::byte_utils;
+use std::collections::VecDeque;
+
+pub(super) type SymmetricKey = [u8; 32];
+
+/// Maximum Lightning message data length according to
+/// [BOLT-8](https://github.com/lightningnetwork/lightning-rfc/blob/v1.0/08-transport.md#lightning-message-specification)
+/// and [BOLT-1](https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md#lightning-message-format):
+const LN_MAX_MSG_LEN: usize = 65535;
+const LN_MAX_PACKET_LENGTH: usize = MESSAGE_LENGTH_HEADER_SIZE + chacha::TAG_SIZE + LN_MAX_MSG_LEN + chacha::TAG_SIZE;
+
+const MESSAGE_LENGTH_HEADER_SIZE: usize = 2;
+const TAGGED_MESSAGE_LENGTH_HEADER_SIZE: usize = MESSAGE_LENGTH_HEADER_SIZE + chacha::TAG_SIZE;
+
+const KEY_ROTATION_INDEX: u32 = 1000;
+
+/// Instantiate a new (Encryptor, Decryptor) with specified sending and receiving keys
+pub fn create_encryptor_decryptor(sending_key: SymmetricKey, receiving_key: SymmetricKey, chaining_key: SymmetricKey) -> (Encryptor, Decryptor) {
```

Comment:
This was just to preserve the original design choice of the handshake code instantiating the Conduit.

The more this review goes on I'm realizing that the assumptions I had about #494 being reviewed/approved at least from a design perspective were naive. I spent a lot of time trying to keep the original design choices but likely should have just started over from the beginning.

I think that pattern makes a lot of sense. The handshake code will just export the keys and the Transport layer will instantiate an Encryptor/Decryptor with them.

I'll put that patch together in the next day or so.

Reply:
This seemed a bit too obvious when I first read the feedback, but I couldn't put my finger on it. I went back through my stack to understand what happened and I wanted to share what I found for the next person that may fix this up.

The original design choice in 494 was to have the handshake code consume all bytes from the user unconditionally. In the event that the handshake was completed, but there was additional data, the bytes were transferred into the Decryptor (inside the handshake code) for later retrieval by the PeerManager. At that point in time, the PeerManager worked with the handshake code directly.

In order to move the Decryptor creation outside the handshake code, the likely design choice would be to have the handshake code return the number of bytes consumed and have the higher layers drain the input buffer. Then, when the handshake completes, it can instantiate the Decryptor with whatever bytes are left.

As the refactoring took place, it made sense to keep this design choice to limit the number of external changes to the code in order to reduce the feedback surface.

I had one outstanding patch that attempted to do this in the pre-Transport layering. Unfortunately, the code inside PeerManager was already very complicated and adding an additional "bytes read" stage to it without tests was too much complexity at the time. So, the design choice stayed.

However, now that there is more distinct layering this is definitely worth looking into again. The PeerManager code would still unconditionally hand off all bytes to the Transport layer and the Transport layer would handle the "bytes read" logic as well as the construction of the Decryptor.

Likely, this will be a much cleaner follow-on PR. This also makes sense since it doesn't really affect the PeerManager interface and is more of an implementation detail of the Transport<->Handshake code relationship.